File: core/data/item_levels.py
"""Level-up requirements for equippable items."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Iterable, Tuple


@dataclass(frozen=True)
class ItemLevelRequirement:
    """Materials and duplicate costs for reaching a specific item level."""

    level: int
    item_cost: int
    materials: Dict[str, int]


# Requirements are defined for the *target* level (e.g. entry with level=2
# represents the materials required to go from level 1 -> level 2).
ITEM_LEVEL_TABLE: Dict[str, Tuple[ItemLevelRequirement, ...]] = {
    "kingdom_key": (
        ItemLevelRequirement(level=2, item_cost=1, materials={"bright_shard": 1}),
        ItemLevelRequirement(
            level=3,
            item_cost=1,
            materials={"bright_shard": 2, "dark_shard": 1},
        ),
        ItemLevelRequirement(
            level=4,
            item_cost=2,
            materials={"bright_shard": 2, "dark_shard": 2},
        ),
        ItemLevelRequirement(
            level=5,
            item_cost=2,
            materials={"bright_shard": 3, "dark_shard": 2, "mythril_fragment": 1},
        ),
    ),
    "mages_staff": (
        ItemLevelRequirement(level=2, item_cost=1, materials={"bright_shard": 1}),
        ItemLevelRequirement(
            level=3,
            item_cost=1,
            materials={"bright_shard": 1, "dark_shard": 1, "mythril_fragment": 1},
        ),
        ItemLevelRequirement(
            level=4,
            item_cost=2,
            materials={"bright_shard": 2, "dark_shard": 2, "mythril_fragment": 1},
        ),
        ItemLevelRequirement(
            level=5,
            item_cost=2,
            materials={"bright_shard": 2, "dark_shard": 2, "mythril_fragment": 2},
        ),
    ),
    "knights_shield": (
        ItemLevelRequirement(level=2, item_cost=1, materials={"dark_shard": 1}),
        ItemLevelRequirement(
            level=3,
            item_cost=1,
            materials={"dark_shard": 2, "mythril_fragment": 1},
        ),
        ItemLevelRequirement(
            level=4,
            item_cost=2,
            materials={"dark_shard": 3, "mythril_fragment": 1},
        ),
        ItemLevelRequirement(
            level=5,
            item_cost=2,
            materials={"dark_shard": 3, "mythril_fragment": 2},
        ),
    ),
    "champion_belt": (
        ItemLevelRequirement(level=2, item_cost=1, materials={"mythril_fragment": 1}),
        ItemLevelRequirement(
            level=3,
            item_cost=1,
            materials={"dark_shard": 1, "mythril_fragment": 2},
        ),
        ItemLevelRequirement(
            level=4,
            item_cost=2,
            materials={"dark_shard": 1, "mythril_fragment": 3},
        ),
        ItemLevelRequirement(
            level=5,
            item_cost=2,
            materials={"dark_shard": 2, "mythril_fragment": 3},
        ),
    ),
    "heros_crest": (
        ItemLevelRequirement(level=2, item_cost=1, materials={"bright_shard": 1}),
        ItemLevelRequirement(
            level=3,
            item_cost=1,
            materials={"bright_shard": 2, "mythril_fragment": 1},
        ),
        ItemLevelRequirement(
            level=4,
            item_cost=2,
            materials={"bright_shard": 2, "mythril_fragment": 2},
        ),
        ItemLevelRequirement(
            level=5,
            item_cost=2,
            materials={"bright_shard": 3, "mythril_fragment": 2},
        ),
    ),
    "elven_bandana": (
        ItemLevelRequirement(level=2, item_cost=1, materials={"bright_shard": 1}),
        ItemLevelRequirement(
            level=3,
            item_cost=1,
            materials={"bright_shard": 1, "dark_shard": 1},
        ),
        ItemLevelRequirement(
            level=4,
            item_cost=2,
            materials={"bright_shard": 2, "dark_shard": 1},
        ),
        ItemLevelRequirement(
            level=5,
            item_cost=2,
            materials={"bright_shard": 2, "dark_shard": 2},
        ),
    ),
}


def iter_level_requirements(item_id: str) -> Iterable[ItemLevelRequirement]:
    """Yield the configured level requirements for an item."""

    return ITEM_LEVEL_TABLE.get(item_id, ())


def requirement_for_level(
    item_id: str,
    target_level: int,
) -> ItemLevelRequirement | None:
    requirements = ITEM_LEVEL_TABLE.get(item_id)
    if not requirements:
        return None
    for requirement in requirements:
        if requirement.level == target_level:
            return requirement
    return None


def max_level(item_id: str) -> int:
    requirements = ITEM_LEVEL_TABLE.get(item_id)
    if not requirements:
        return 1
    return requirements[-1].level


__all__ = [
    "ItemLevelRequirement",
    "ITEM_LEVEL_TABLE",
    "iter_level_requirements",
    "requirement_for_level",
    "max_level",
]

File: web/src/domain/item_levels.ts
export type ItemLevelRequirement = {
  level: number;
  itemCost: number;
  materials: Record<string, number>;
};

const ITEM_LEVEL_TABLE: Record<string, ItemLevelRequirement[]> = {
  kingdom_key: [
    { level: 2, itemCost: 1, materials: { bright_shard: 1 } },
    { level: 3, itemCost: 1, materials: { bright_shard: 2, dark_shard: 1 } },
    { level: 4, itemCost: 2, materials: { bright_shard: 2, dark_shard: 2 } },
    {
      level: 5,
      itemCost: 2,
      materials: { bright_shard: 3, dark_shard: 2, mythril_fragment: 1 }
    }
  ],
  mages_staff: [
    { level: 2, itemCost: 1, materials: { bright_shard: 1 } },
    {
      level: 3,
      itemCost: 1,
      materials: { bright_shard: 1, dark_shard: 1, mythril_fragment: 1 }
    },
    {
      level: 4,
      itemCost: 2,
      materials: { bright_shard: 2, dark_shard: 2, mythril_fragment: 1 }
    },
    {
      level: 5,
      itemCost: 2,
      materials: { bright_shard: 2, dark_shard: 2, mythril_fragment: 2 }
    }
  ],
  knights_shield: [
    { level: 2, itemCost: 1, materials: { dark_shard: 1 } },
    {
      level: 3,
      itemCost: 1,
      materials: { dark_shard: 2, mythril_fragment: 1 }
    },
    {
      level: 4,
      itemCost: 2,
      materials: { dark_shard: 3, mythril_fragment: 1 }
    },
    {
      level: 5,
      itemCost: 2,
      materials: { dark_shard: 3, mythril_fragment: 2 }
    }
  ],
  champion_belt: [
    { level: 2, itemCost: 1, materials: { mythril_fragment: 1 } },
    {
      level: 3,
      itemCost: 1,
      materials: { dark_shard: 1, mythril_fragment: 2 }
    },
    {
      level: 4,
      itemCost: 2,
      materials: { dark_shard: 1, mythril_fragment: 3 }
    },
    {
      level: 5,
      itemCost: 2,
      materials: { dark_shard: 2, mythril_fragment: 3 }
    }
  ],
  heros_crest: [
    { level: 2, itemCost: 1, materials: { bright_shard: 1 } },
    {
      level: 3,
      itemCost: 1,
      materials: { bright_shard: 2, mythril_fragment: 1 }
    },
    {
      level: 4,
      itemCost: 2,
      materials: { bright_shard: 2, mythril_fragment: 2 }
    },
    {
      level: 5,
      itemCost: 2,
      materials: { bright_shard: 3, mythril_fragment: 2 }
    }
  ],
  elven_bandana: [
    { level: 2, itemCost: 1, materials: { bright_shard: 1 } },
    {
      level: 3,
      itemCost: 1,
      materials: { bright_shard: 1, dark_shard: 1 }
    },
    {
      level: 4,
      itemCost: 2,
      materials: { bright_shard: 2, dark_shard: 1 }
    },
    {
      level: 5,
      itemCost: 2,
      materials: { bright_shard: 2, dark_shard: 2 }
    }
  ]
};

export function iterLevelRequirements(itemId: string): ItemLevelRequirement[] {
  return ITEM_LEVEL_TABLE[itemId] ?? [];
}

export function requirementForLevel(
  itemId: string,
  targetLevel: number
): ItemLevelRequirement | null {
  const requirements = ITEM_LEVEL_TABLE[itemId];
  if (!requirements) return null;
  const match = requirements.find((req) => req.level === targetLevel);
  return match ?? null;
}

export function maxLevel(itemId: string): number {
  const requirements = ITEM_LEVEL_TABLE[itemId];
  if (!requirements || requirements.length === 0) return 1;
  return requirements[requirements.length - 1].level;
}

export { ITEM_LEVEL_TABLE };
